paddle/math/Matrix.cpp

```diff
@@ -202,6 +202,17 @@ void GpuMatrix::resetOne() {
   CHECK(data_ != NULL);
   one();
 }
+
+void GpuMatrix::setDiag(real value) {
+  CHECK(data_ != NULL);
+  CHECK_EQ(height_, width_);
+
+  zeroMem();
+  for (size_t i = 0; i < height_; i++) {
```

Comment:
Use BaseMatrix::assign() for this will be more efficient. This can avoid multiple hl_memcpy, which is very inefficient:
BaseMatrix diag(height_, 1, stride_ + 1, data_, false, useGpu_);
diag.assign(value);

And this can be implemented as as method in BaseMatrix, which works for both cpu and gpu.